src/main/java/jenkins/plugins/openstack/compute/JCloudsPreCreationThread.java

```diff
@@ -0,0 +1,92 @@
+package jenkins.plugins.openstack.compute;
+
+import java.lang.Math;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+import hudson.Extension;
+import hudson.model.TaskListener;
+import hudson.model.AsyncPeriodicWork;
+
+/**
+ * Periodically ensure enough slaves are created.
+ */
+@Extension @Restricted(NoExternalUse.class)
+public final class JCloudsPreCreationThread extends AsyncPeriodicWork {
+    private static final Logger LOGGER = Logger.getLogger(JCloudsPreCreationThread.class.getName());
+
+    public JCloudsPreCreationThread() {
+        super("OpenStack slave pre-creation");
+    }
+
+    @Override
+    public long getRecurrencePeriod() {
+        return MIN * 2;
+    }
+
+    @Override
+    public void execute(TaskListener listener) {
+        for (JCloudsCloud cloud : JCloudsCloud.getClouds()) {
+            for (JCloudsSlaveTemplate template : cloud.getTemplates()) {
+                SlaveOptions slaveOptions = template.getEffectiveSlaveOptions();
+                int instancesMin = slaveOptions.getInstancesMin();
+                int retentionTime = slaveOptions.getRetentionTime();
+                if (instancesMin > 0) {
+                    int globalMaxInstances = cloud.getEffectiveSlaveOptions().getInstanceCap();
+                    int templateMaxInstances = slaveOptions.getInstanceCap();
+                    int maxNodes = Math.min(templateMaxInstances, globalMaxInstances);
+                    // If retentionTime==0, take this as an indication that "used" instances should not
+                    // be reused and thus do not count them as running instances.
+                    int runningNodeTotal = template.getActiveNodesTotal(retentionTime == 0);
```

Comment:
NIT: The information about the retention time would be better extracted from the node's `SlaveOptions` as template configuration might have changed since the node was provisioned.

Reply:
Actually, here I'm iterating over the templates and pre-provisioning images based on the template configuration; if the config of a particular node has diverged, it shouldn't influence the pre-provisioning of instances. Do you agree?

Reply:
The only way for template and node to diverge in their options is the template changes. And since the retention strategy take the timeout from the node and not its template, the 2 async tasks can in theory work with deferent timeout values. We should stick with the the one or the other.

The was I see this is the nodes are products of a template that might not exists or maintain the same config over the lifetime of a node and that is why they copy the options so the config is consistent (and always present) for the node's life.

Reply:
I totally agree but in the context of `JCloudsPreCreationThread.execute` we don't deal with the nodes: we simply iterate over the templates.

Now, if `retentionTime` diverges (between template and nodes) I argue that nothing bad can happen: if the retention time differs then the worse that can happen is the situation where the new retention time on the template is now != zero and it's zero on an already running node: the result is that we could pre-provision one additional instance more aggressively (but still within the bounds of the max instance cap).